File: solutions/python/crypto-square/1/crypto_square.py
import re
import math


def cipher_text(plain_text):
    text = re.sub(r'[^a-z0-9]', '', plain_text.lower())

    if not text:
        return ""

    length = len(text)

    c = math.ceil(math.sqrt(length))
    r = math.ceil(length / c)

    text += " " * (c * r - length)

    rows = [text[i:i + c] for i in range(0, len(text), c)]

    result = []
    for i in range(c):
        col = "".join(row[i] for row in rows)
        result.append(col)

    return " ".join(result)

File: solutions/cpp/crypto-square/1/crypto_square.cpp
#include "crypto_square.h"

#include <algorithm>
#include <cmath>

namespace crypto_square {
    cipher::cipher(std::string plaintext) {
        plaintext.erase(
            std::remove_if(plaintext.begin(), plaintext.end(), [](const unsigned char c) {
                return !std::isalnum(c);
            }),
            plaintext.end()
        );
        if (plaintext.empty()) {
            encode.append("");
            return;
        }
        const int columns = static_cast<int>(std::ceil(sqrt(static_cast<double>(plaintext.size()))));
        const int rows = static_cast<int>(std::ceil(static_cast<double>(plaintext.size())/columns ));
        const int result_size = rows * columns + (columns - 1) ;
        int pos = 0;
        encode.reserve(result_size);
        for (int i = 0; i < result_size; i++) {
            encode.push_back(' ');
        }
        for (int col = 0; col < columns; col++) {
            if (col > 0) {
                pos++;
            }
            for (int row = 0; row < rows; row++) {
                if (const int index = row * columns + col; index < static_cast<int>(plaintext.size())) {
                    encode[pos++] = static_cast<char>(std::tolower(plaintext[index]));
                }else {
                    pos++;
                }
            }
        }
    }

    std::string cipher::normalized_cipher_text() {
        return encode;
    }

} // namespace crypto_square
